### PR TITLE
Implement GraphQL Rate Limiting

### LINKED ISSUE #2777: Implement rate limiting of GraphQL mutations
We should rate limit our GraphQL mutations. Apart from the changes in #2758, mutations should have general rate limits configured in the schema to avoid excessive automatic spam from bots who call our API 20 times a second.

Those could look something like this:

```GraphQL
type Mutation {
  addMessage: Message        @rateLimit(perMinute: 20)
  publishThread: Thread      @rateLimit(perMinute: 5, perHour: 100)
  createCommunity: Community @rateLimit(perMinute: 5)
  createChannel: Community   @rateLimit(perMinute: 5)
  createDMThread: DMThread   @rateLimit(perMinute: 5)
}
```

> Note: not sure about that API just yet, gotta bikeshed it.

- Are there any packages that implement this already? Couldn't find any
- What should the API look like?
- Where do we store the counts? Redis
- There's gotta be existing non-GraphQL-specific solutions, how do those work?

### LINKED ISSUE #3901: Rate limit dm creation
Add rate limiting so one person can't spam lots of people with new dm threads

### PR DESCRIPTION
**Pull request by mxstbr**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api

**Related issues**
Closes #3901, closes #2777, supersedes #2874

~~This doesn't work, currently it doesn't rate limit at all, not even the `identifyContext` option of the directive is called :confused: /cc @kirkness any ideas what I'm doing wrong?~~ It works!

Adds rate limiting to our `createX` GraphQL mutations (tbd what else needs rate limiting). The rate limits are shared between all API instances, stored in Redis, and we identify a specific user by either their authenticated user ID or their request IP if they are unauthenticated.

### CODE REVIEW

**Comment by kirkness** [comment]
Yo @mxstbr! Will check this out now, setup looks good to me though. Have you tried the InMemoryStore, just to reduce the problem?

Any tips on best ways to test this locally?

**Comment by mxstbr** [comment]
Have tried InMemoryStore too, same issue.

Best way to test this locally would probably be to go through the setup steps in the readme (should only take a couple minutes max) and then adding a console.log to the identifyContext function while the app is running. You'll notice nothing happens

**Comment by kirkness** [comment]
Ok got it working, you just need to move the `schemaDirectives` config from your `ProtectedApolloServer` config to your `makeExecutableSchema` config. i.e.

```js
makeExecutableSchema({
  typeDefs: [],
  resolvers,
  schemaDirectives: {
    rateLimit: rateLimitDirective,
  },
})
```
Not sure why this is the case as in my example I used ApolloServer.

**Comment by mxstbr** [comment]
Amazing, thanks so so much! :100:

**Comment by mxstbr** [comment]
@kirkness does the Redis store expire records automatically or do they stick around?

**Comment by kirkness** [comment]
@mxstbr they currently just stick around.

**Comment by mxstbr** [comment]
Thanks so much for the help @kirkness, this now works beautifully! 😍 

e2e test failures seem unrelated, as there is not a single GraphQL error log in the API logs. Other than that this is ready for a review @brianlovin!

**Comment by kirkness** [comment]
@mxstbr - amazing, happy to help. I'm a huge fan of Spectrum - glad you find it helpful.

**Comment by brianlovin** [comment]
@mxstbr I'll re-run the CI workflow, but ideally all e2e tests pass before we merge this :)

**Comment by mxstbr** [comment]
There ya go 😆

**Comment by mxstbr** [comment]
Maybe we should turn rate limiting off during testing? I can tackle that tomorrow!

**Comment by brianlovin** [comment]
That'd be great! In the meantime, this lgtm

**Comment by kirkness** [comment]
Will get [this](https://github.com/teamplanes/graphql-rate-limit/issues/7) implemented. 

Also FYI: 
https://github.com/teamplanes/graphql-rate-limit/pull/16
https://github.com/teamplanes/graphql-rate-limit/pull/15

File: api/types/Message.js [hunk 1]
```diff
@@ -58,6 +58,7 @@ const Message = /* GraphQL */ `
 
   extend type Mutation {
     addMessage(message: MessageInput!): Message
+      @rateLimit(max: 20, window: "1m")
```
Comment (brianlovin) [inline api/types/Message.js]: This might actually be too low. For a fast conversation, wdyt? Maybe it should be 30?

Comment (mxstbr) [reply api/types/Message.js]: 30 per minute? Okay

File: api/utils/rate-limit-directive.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import { createRateLimitDirective, RedisStore } from 'graphql-rate-limit';
+import { getClientIp } from 'request-ip';
+import createRedis from 'shared/bull/create-redis';
+import ms from 'ms';
+
+export default createRateLimitDirective({
+  identifyContext: ctx => (ctx.user && ctx.user.id) || getClientIp(ctx.request),
+  store: new RedisStore(createRedis()),
+  formatError: ({ fieldName, fieldIdentity, max, window }) =>
+    `Slow down there cowboy! You've called '${fieldName}' ${max} times in the past ${ms(
```
Comment (brianlovin) [inline api/utils/rate-limit-directive.js]: We could say "Slow down there partner!" to have a neutral gender

File: api/types/Message.js (hunk 1 above)
Comment (brianlovin) [reply api/types/Message.js]: I don't know...i mean 20 seems high, I'm just imaging a text thread where things are going fast. 20 is probably reasonable though.

**Comment by mxstbr** [comment]
Done all those things! :100: Ready for another review

**Review by brianlovin — APPROVED** [review]

**Pull request merged**